lib/Widgets/HyperTextView.vala

```diff
@@ -0,0 +1,317 @@
+
+    /**
+     * This class enables navigatable URLs in Gtk.TextView
+     */
+    public class HyperTextView : Gtk.TextView {
+
+        private uint buffer_changed_debounce_timeout_id = 0;
+        private int buffer_cursor_position_when_change_started = 0;
+
+        private GLib.HashTable<string, Gtk.TextTag> uri_text_tags;
+        private Regex uri_regex;
+
+        private bool is_control_key_pressed = false;
+
+        construct {
+            var http_charset = "[a-zA-Z0-9_\\/\\-\\+\\.:@\\?&%=#]";
+            var email_charset = "[a-zA-Z0-9_\\-\\.]";
+            var email_tld_charset = "[a-zA-Z0-9_\\-]";
+
+            var http_match_str = "https?:\\/\\/" + http_charset + "{1,}";
+            var email_match_str = "(mailto:)?" + email_charset + "{1,}@" + email_charset + "{1,}\\." + email_tld_charset + "{1,}";
+
+            var uri_regex_str = "(?:(" +
+                    http_match_str +
+                ")|(" +
+                    email_match_str +
+                "))";
+
+            uri_text_tags = new GLib.HashTable<string, Gtk.TextTag> (str_hash, direct_equal);
+            try {
+                uri_regex = new Regex (uri_regex_str);
+            } catch (GLib.RegexError e) {
+                critical ("RegexError while constructing URI regex: %s", e.message);
+            }
+
+            buffer.notify["cursor-position"].connect (on_buffer_cursor_position_changed);
+            buffer.paste_done.connect (on_paste_done);
+            buffer.changed.connect_after (on_after_buffer_changed);
+
+            button_release_event.connect (on_button_release_event);
+            motion_notify_event.connect (on_motion_notify_event);
+            focus_out_event.connect (on_focus_out_event);
+
+            /**
+             * Binding key_press/key_release signals to toplevel window
+             * if possible enables us to detect when the Control key
+             * is pressed even when HyperTextView is not focused.
+             */
+
+            Gtk.Window? toplevel_window = null;
+            foreach (unowned var window in Gtk.Window.list_toplevels ()) {
+                if (window.get_parent_window () == null) {
+                    toplevel_window = window;
+                    break;
+                }
+            }
+
+            if (toplevel_window != null) {
+                toplevel_window.key_press_event.connect (on_key_press_event);
+                toplevel_window.key_release_event.connect (on_key_release_event);
+            } else {
+                warning ("Unable to bind key press events to toplevel window, Control + Click may not behave correct under all circumstances.");
+                // bind to this as a fallback
+                key_press_event.connect (on_key_press_event);
+                key_release_event.connect (on_key_release_event);
+            }
+        }
+
+        private void on_buffer_cursor_position_changed () {
+            if (buffer_cursor_position_when_change_started == 0) {
```

Comment:
Shouldnt this be `if (buffer_cursor_position_when_change_started == -1` (when this variable is initialized to `-1`)?

Otherwise the value of this variable cannot remain `0`